### PR TITLE
feat(scan): opt-in location_filter.strict fails closed on empty locations (#3276)

### LINKED ISSUE #3276: location_filter fails open on empty locations, which defeats location-restricted scan-ats-full sweeps (iCIMS)
## Summary

`buildLocationFilter` passes any posting whose location **and** URL hint are both empty, even when `location_filter.allow` is non-empty. For providers that do not reliably return a location, this silently inverts the filter: instead of "only these places", the user gets "everything from that provider, plus matches from everywhere else".

`scan-ats-full.mjs` imports the same filter (`scan-ats-full.mjs:51`) and applies it at lines 436, 755 and 765, so a reverse-ATS sweep restricted to one region returns almost entirely out-of-region results.

## The line

`scan.mjs:351`

```js
// Nothing to judge on either field → pass (don't penalize missing data).
if (lower === '' && hint === '') return true;
```

## Reproduction

```js
import { buildLocationFilter } from './scan.mjs';

const f = buildLocationFilter({ allow: ['Puerto Rico', 'San Juan'] });

f('',             'https://careers-peraton.icims.com/jobs/168257/data-architect/job'); // true  ← unexpected
f('',             'https://careers-here.icims.com/jobs/81523/senior-software-engineer/job'); // true  ← unexpected
f('San Juan, PR', 'https://example.com/j'); // true  ← correct
f('Boston, MA',   'https://example.com/j'); // false ← correct
```

The filter behaves correctly whenever a location string exists. The failure is confined to the empty-location path.

## Observed impact

A single `scan-ats-full.mjs --since 45 --dry-run` run with `location_filter.allow` restricted to one region (no `always_allow`, no bare-country entries) produced **59 matches, 0 of them in the requested region**:

| Location reported | Matches |
|---|---:|
| `N/A` (empty) | **51** |
| `United States (Remote)` | 2 |
| `Remote (Anywhere in the world)` | 1 |
| `Remote - United States` | 1 |
| `EMEA(Remote)` | 1 |
| `New York` | 1 |
| `Lehi, Utah` | 1 |
| `Boston, MA` | 1 |

All 51 empty-location rows came from `icims-full`. iCIMS URLs (`careers-{tenant}.icims.com/jobs/{id}/{slug}/job`) carry no geography, so `locationHintFromUrl` returns `''` too and the pass short-circuit fires for every one of them.

Note `providers/icims.mjs:86-88` already documents the *opposite* failure — a selector bug that read an empty location off some tenants and caused postings to be **dropped**. The regex was fixed, but empty locations still occur in practice, and now they fail open instead of closed.

## Why the current behavior is defensible in one context and wrong in another

For `scan.mjs` over a curated `tracked_companies` list, "don't penalize missing data" is reasonable: the user chose those employers, so a posting with no location is worth surfacing.

For `scan-ats-full.mjs`, the company list is the entire public ATS dataset. The user never opted into those employers, and `location_filter` is the *only* geographic constraint in the run. Failing open there removes the sole filter that made the sweep meaningful.

A user cannot tell this happened without inspecting the location column of every result. The run reports matches and looks successful.

## Possible directions

Not prescribing a specific fix, but options that seem consistent with the existing design:

1. **Make the empty-location policy explicit and configurable**, e.g. `location_filter.on_missing: pass | reject` (default `pass`, preserving current behavior), and have `scan-ats-full.mjs` default it to `reject`.
2. **Scope the change to the reverse scanner only** — pass a flag into `buildLocationFilter` so the full-dataset sweep rejects unknown locations while `scan.mjs` is untouched.
3. **Report it rather than change it** — count empty-location passes and print them as a distinct line in the run summary (`N matches passed with no location data`), the way `droppedNoDate` and `droppedContent` are already surfaced. This alone would have made the problem obvious immediately.

(3) is the smallest change and would remove the silence, which is the actual harm here. (1) or (2) would fix the filtering.

## Environment

- career-ops v1.28.0
- Node v22.22.3, macOS (darwin 25.5.0)
- Reproduced with `scan-ats-full.mjs --since 45 --dry-run --verbose`, `--ats` unrestricted

### PR DESCRIPTION
**Pull request by Zoubeir23**
Closes #3276 (`confirmed`, design specified by @santifer).

## The bug

`buildLocationFilter()` short-circuits to **pass** when a posting has no location at all — empty structured field **and** no city in the URL — before the `allow` list is ever consulted:

```js
if (lower === '' && hint === '') return true;   // "don't penalize missing data"
```

For a broad sweep that default is right. For a **location-restricted** sweep over a provider that never returns a structured location — **iCIMS** is the common case, its URLs (`careers-{tenant}.icims.com/jobs/{id}/{slug}/job`) carry no geography either — it silently inverts the filter: the "only these places" `allow` list is never applied, and every out-of-region posting passes.

From the issue, one `scan-ats-full.mjs --since 45 --dry-run` run restricted to one region: **59 matches, 0 in the requested region**, 51 of them empty-location iCIMS rows.

@santifer:
> The fix that respects both users: an opt-in `location_filter.strict: true` that fails closed on empty fields, default unchanged.

## The fix

`location_filter.strict: true` (opt-in). When set **and** a restricting tier (`allow` / `block` / `block_hard`) is configured, a location-less posting is **rejected** instead of passed. `strict: true` alone restricts nothing and stays inert — the flag only bites when there is a restriction to enforce.

- **`scan.mjs`** — the `strict` flag plus the single gated `return`:
  ```js
  if (lower === '' && hint === '') return !strict;
  ```
- **`validate-portals.mjs`** — type-checks `strict` (`must be a boolean when set`), same shape as `visa_filter.enabled`.
- **`templates/portals.example.yml`** — documents when to turn it on (iCIMS-class providers, restricted sweeps).

## Tests — `test-all.mjs` cases 28–29

- **28:** the iCIMS repro from the issue — empty location + non-empty `allow` passes by default, is rejected under `strict`; a posting that *does* carry a matching location still passes; a real out-of-region location still rejects.
- **29:** `strict: true` with no `allow`/`block`/`block_hard` is inert (passes everything); a block-only `strict` config fails closed on empty locations but passes a real non-blocked one.

## Test plan

- [x] `node validate-portals.mjs --self-test`
- [x] `node validate-portals.mjs --file templates/portals.example.yml` — 0 errors
- [x] `node --test tests/providers/icims.test.mjs tests/scan-dedup-location-aware.test.mjs`
- [x] `node test-all.mjs` — 8689 passed, 0 failed (16 pre-existing warnings: external board 503s, Workday truncation notes)

## Summary

Users can set `location_filter.strict: true` to reject postings with no location data when `allow`, `block`, or `block_hard` is configured. Matching locations still pass, and out-of-region locations remain rejected. Default behavior remains unchanged. Strict mode alone has no effect. Non-boolean values fail validation.

## Changes

- `scan.mjs:287-296, 519-549`: Adds strict handling for empty locations.
- `validate-portals.mjs:165-174`: Requires `strict` to be boolean.
- `templates/portals.example.yml:103-120`: Documents strict mode and `block_hard`.
- `modes/scan.md:218-221`: Documents filtering behavior.
- `tests/location-filter-strict.test.mjs:5-36`: Covers empty locations and strict edge cases.
- `tests/validate-portals-location-filter-strict.test.mjs:24-53`: Covers valid, invalid, and omitted values.
- `test-all.mjs:7775`: Points to the dedicated strict-filter tests.

System files touched: `modes/` is touched through `modes/scan.md`. `AGENTS.md`, `update-system.mjs`, `DATA_CONTRACT.md`, `providers/`, and `.github/` are unchanged.

### CODE REVIEW

File: test-all.mjs [hunk 1]
```diff
@@ -7772,6 +7772,50 @@ try {
     fail('remote-title rescue changed behavior for non-remote or malformed titles');
   }
 
+  // Case 28: `location_filter.strict` (#3276). A location-restricted sweep over
```
[the remark this answers is not in the document]
Comment (Zoubeir23) [reply test-all.mjs]: Done. Cherry-picked the split @Quantropy pushed to `chipoto69:steward/pr4033-reviewfix-20260909` (`07e1aa7`, credit preserved): the `location_filter.strict` assertions now live in `tests/location-filter-strict.test.mjs` as a `node:test` suite, and `test-all.mjs` keeps only a one-line pointer. `node --test tests/location-filter-strict.test.mjs` → 2/2, full `test-all` → 8688 passed, 0 failed.

File: templates/portals.example.yml [hunk 2]
```diff
@@ -100,6 +100,15 @@
...
 #
+# `strict: true` (optional) fails CLOSED when a posting has no location at all
+# (empty structured field AND no city in the URL). The default passes such a
+# posting — "don't penalize missing provider data" — which is right for a
+# broad sweep. But a location-*restricted* sweep over a provider that never
+# returns a location (iCIMS is the common case) then lets every out-of-region
+# posting through: the "only these places" allow list is never consulted.
+# Turn strict on for those sweeps. It only bites when a restricting tier
+# (allow / block / block_hard) is set; `strict: true` alone restricts nothing.
+#
```
[the remark this answers is not in the document]
Comment (Zoubeir23) [reply templates/portals.example.yml:111]: Skipping this one: `career-ops-docs` is a separate repository, not present in this checkout — there is no `faq.mdx` (or any `*.mdx`) here to patch. The steward's own bounded pass confirmed the same ("this checkout contains no `*.mdx` files"). The `templates/portals.example.yml` block added in this PR is the discoverable in-repo documentation; a `career-ops-docs` FAQ update belongs in that repo's own PR.

**Review by Scott-Emberson — COMMENTED** [review]
Verified the owned test by execution and mutation. It is sound.

`tests/location-filter-strict.test.mjs` imports the real `buildLocationFilter` from `scan.mjs` and drives it over fixtures, no hand-reconstructed filter. Both directions are pinned: with strict on, an unconfirmable location fails closed (an empty location against an iCIMS URL, an empty location with no URL, a null/null pair all return false), and with strict off, the default behavior is unchanged (the same empty-location-plus-URL case passes), alongside positive and negative controls. I mutation-checked three ways: fail-open under strict reddens, always-fail-closed reddens (which proves the default-off path is genuinely asserted, not just the strict path), and dropping the tier gate reddens the "strict with no restricting tier is inert" assertion. Reverting restores green.

One in-lane note: the PR also adds a `location_filter.strict must be a boolean` check in `validate-portals.mjs`, but the owned test imports only `buildLocationFilter`, so that config-validation branch is not covered here. It is a small gap, not a defect in the filter test, and worth one added case if you want the validation pinned too.

Sound on the test axis. `scan.mjs` and `validate-portals.mjs` are the maintainer's call on the routing.

**Review by Scott-Emberson — COMMENTED** [review]
Re-verified at this head, and the coverage gap I raised is now closed. The three original files (the owned test, scan.mjs, validate-portals.mjs) are byte-identical to my previous review, so that verdict carries, and the new commit is purely additive: it adds `tests/validate-portals-location-filter-strict.test.mjs`, which drives the real validate-portals.mjs CLI over three cases (strict true accepted, strict "yes" rejected by name, strict omitted stays valid). That covers the `location_filter.strict must be a boolean` check I noted was previously unpinned. Both suites pass (2 plus 3), and neutering the boolean check to fail-open reddens the reject-a-non-boolean assertion.

Owned tests sound; branch behind main. SIGNAL (scan-core), so not surfacing as ready-to-approve; routing and merge are the maintainer's.

File: templates/portals.example.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Zoubeir23) [reply templates/portals.example.yml:111]: No follow-up issue needed — `templates/portals.example.yml:103-111` is the in-repo documentation this PR is responsible for, and it's in place. A `career-ops-docs` FAQ update is out of scope for this checkout; leaving that to whoever maintains that repo.

**Review by santifer — APPROVED** [review]

**Comment by santifer** [comment]
Merged. Thank you, @Zoubeir23: your change is now in `main`.

If this was your first contribution here, welcome aboard. If you want to keep going, `good first issue` and `help wanted` are curated for exactly that, and you can ask on any of them before starting.
